src/griffe/git.py

```diff
@@ -0,0 +1,120 @@
+"""This module contains the code allowing to load modules from specific git commits.
+
+```python
+from griffe.git import load_git
+
+# where `repo` is the folder *containing* `.git`
+old_api = load_git("my_module", commit="v0.1.0", repo="path/to/repo")
+```
+"""
+from __future__ import annotations
+
+import os
+import shutil
+from contextlib import contextmanager
+from pathlib import Path
+from subprocess import DEVNULL, PIPE, CalledProcessError, check_output, run  # noqa: S404
+from tempfile import TemporaryDirectory
+from typing import Any, Iterator, Sequence
+from uuid import uuid1
+
+from griffe import loader
+from griffe.agents.extensions import Extensions
+from griffe.collections import LinesCollection, ModulesCollection
+from griffe.dataclasses import Module
+from griffe.docstrings.parsers import Parser
+
+
+def _assert_git_repo(repo: str) -> None:
+    if not shutil.which("git"):
+        raise RuntimeError("Could not find git executable. Please install git.")
+
+    try:
+        check_output(["git", "-C", repo, "rev-parse", "--is-inside-work-tree"], stderr=DEVNULL)  # noqa: S603,S607
+    except CalledProcessError as err:
+        raise OSError(f"Not a git repository: {repo!r}") from err
+
+
+@contextmanager
+def tmp_worktree(commit: str = "HEAD", repo: str | Path = ".") -> Iterator[str]:
+    """Context manager that checks out `commit` in `repo` to a temporary worktree.
+
+    Parameters:
+        commit: A "commit-ish" - such as a hash or tag.
+        repo: Path to the repository (i.e. the directory *containing* the `.git` directory)
+
+    Yields:
+        The path to the temporary worktree.
+
+    Raises:
+        OSError: If `repo` is not a valid `.git` repository
+        RuntimeError: If the `git` executable is unavailable, or if it cannot create a worktree
+    """
+    repo = str(repo)
+    _assert_git_repo(repo)
+    with TemporaryDirectory() as td:
+        uid = f"{str(uuid1())[:5]}_{commit}".replace(" ", "_")
```

Comment:
Is there a reason why you chose UUID1 instead of UUID4 for example? And why you keep only the first 5 characters? Also, I don't see how spaces could be found in the result :confused: UUID1 and commit sha are only composed of `[0-9a-f]`, no?

Reply:
really, there's not much need for a uuid at all :joy: The commit-ish should be sufficient.  I removed uuid and prefixed instead with `griffe_` ... to indicate who made the worktree (on the very little chance that the cleanup after the `finally` fails

Reply:
Great! That simplifies a lot :slightly_smiling_face: Thanks